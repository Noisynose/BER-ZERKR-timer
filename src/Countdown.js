import React, { Component } from 'react';

class Countdown extends Component {
    constructor(props) {
        super(props);

        this.state = { 
            milliseconds : this.props.milliseconds || 0,
            refreshRate : this.props.decrement || 1000,
            status : ((this.props.milliseconds||0) <= 0) ? "FINISHED" : "RUNNING" ,
        };
    }

    componentDidMount() {
        this.timerIdentifier = setInterval(
            () => this.tick(),
            this.state.refreshRate
        );
    }

    componentWillUnmount() {
        this.stop();
    }

    stop() {
        if (this.timerIdentifier) {
            clearInterval(this.timerIdentifier);
        }
    }

    tick() {
        const millisecondsToGo = this.state.milliseconds - this.state.refreshRate;

        this.setState({
            milliseconds : millisecondsToGo,
            status : ( millisecondsToGo <= 0 ) ? "FINISHED" : "RUNNING",
        });

        if (this.state.status === "FINISHED") {
            this.stop();
        }
    }

    render() {
        return (
            <div className="countdown">
                {this.state.milliseconds}
            </div>
        );
    }
}

export default Countdown;