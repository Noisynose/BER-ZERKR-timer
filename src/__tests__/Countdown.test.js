import React from 'react';
import renderer from 'react-test-renderer';
import Countdown from '../Countdown';

jest.useFakeTimers();

describe('Countdown component', () => {
    it('renders without crashing', () => {
        const countdownRender = renderer.create(<Countdown />);
        countdownRender.unmount();
    });

    it('renders correctly', () => {
        const countdownRender = renderer.create(<Countdown />);
        const countdown = countdownRender.toJSON();

        expect(countdown).toMatchSnapshot();
    });

    it('shows up as finished once countdown is done', () => {
        const countdownRender = renderer.create(
            <Countdown milliseconds="1000" />
        );
        const countdownInstance = countdownRender.getInstance();

        expect(countdownInstance.state.status).toBe("RUNNING");

        jest.advanceTimersByTime(500);
        expect(countdownInstance.state.status).toBe("RUNNING");

        jest.runAllTimers();
        expect(countdownInstance.state.status).toBe("FINISHED");

        jest.clearAllTimers();
        countdownRender.unmount();
    });
});