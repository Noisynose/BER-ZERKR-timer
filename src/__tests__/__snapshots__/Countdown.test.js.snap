// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Countdown component renders correctly 1`] = `
<div
  className="countdown"
>
  0
</div>
`;
